feat: support default path environment variable

liweitian opened the pull request:
## Description
Allow user to set default path as an environment variable. If the path is invalid, we still set it to path.join(os.homedir(), 'Documents', 'Composer').
 
Did a merge operation on data.json and data.template.json if it already exists.

## Task Item
Closes #1610 
Closes #1363 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have functionally tested my change

## Screenshots

Please include screenshots or gifs if your PR include UX changes.

a-b-r-o-w-n commented on Composer/packages/server/src/store/store.ts:
```diff
@@ -15,14 +17,24 @@ const dataStorePath =
+    // for some very odd reason on Azure webapp, fs.readFileSync after writeFileSync doesn't
+    // always find the file, add one more io to kick the  virtual file system
+    fs.appendFileSync(dataStorePath, ' ');
+  }
```
We should separate the "migration" and executing the migration.

```ts
let data = initData;

if (fs.existsSync(dataStorePath)) {
  const userData = JSON.parse(fs.readFileSync(dataStorePath, 'utf-8'));
  data = runMigrations(userData);
}

fs.writeFileSync(dataStorePath, JSON.stringify(data, null, 2) + '\n');

if (isHostedInAzure) {
    // for some very odd reason on Azure webapp, fs.readFileSync after writeFileSync doesn't
    // always find the file, add one more io to kick the  virtual file system
    fs.appendFileSync(dataStorePath, ' ');
}
```

```ts
const migrations = [
  ensureDefaultPath,
  ...etc
];

function runMigration(initialData) {
  return migrations.reduce((userData, migration) => {
    return migration(userData);
  }, initialData);
}
```

liweitian commented:
@a-b-r-o-w-n I changed the condition in the migration. The previous would only check whether the data.json contain the defaultPath. But we still need to set it if the defaultPath changed by user.

And I added the following code. 
let folder = process.env.COMPOSER_BOTS_FOLDER;
if (folder && folder.endsWith(':')) {
  folder = folder + '/';
}

The issue is quite odd here: if user set the defaultPath as D: we use fs.existsSync('D:') in many places to check whether the path is valid, and it will return true. However, Path.resolve('D:') will return the current folder D:\BotFramework-Composer\Composer\packages\server>, which means for path module D: is invalid. I feel the fs and path module are not designed consistent. I searched online but cannot find an elegant way to solve this. I checked our code in the FileSelector.tsx we have **itemPath = itemPath[itemPath.length - 1] === ':' ? `${itemPath}/` : itemPath;**
